[Clang] Lift HIPSPV onto the new offload driver

linehill opened the pull request:
Update HIPSPV toolchain to support `--offload-new-driver`. Additionally,  tailor `llvm-spirv` invocation for [chipStar](https://github.com/CHIP-SPV/chipStar) via `spirv64-*-chipstar` offload triple.

Depends on one commit from #170467 and one from #170655.

Commit:
HIPSPV: set allowed SPIR-V extensions for chipStar

linehill committed:
Lift HIPSPV onto the new offload driver

jhuber6 commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -9051,6 +9051,7 @@ void LinkerWrapper::ConstructJob(Compilation &C, const JobAction &JA,
       OPT_v,
       OPT_cuda_path_EQ,
       OPT_rocm_path_EQ,
+      OPT_hip_path_EQ,
```
Is the HIP path really different from the rocm path? Figured those would just be aliases.

linehill replied:
Not necessarily? HIP is a subset of ROCm and I think that  HIP is a subdirectory in ROCm installations so `--hip-path` != `--rocm-path`.

linehill committed:
HIPSPV: Add checks for --offload-new-driver

linehill committed:
HIPSPV: drop --spirv-max-version if subarch is defined

jhuber6 commented on clang/include/clang/Options/Options.td:
```diff
@@ -1003,6 +1003,10 @@ def Xthinlto_distributor_EQ : CommaJoined<["-"], "Xthinlto-distributor=">,
+def Xoffload_compiler : JoinedAndSeparate<["-"], "Xoffload-compiler">,
+  Visibility<[ClangOption, FlangOption]>,
+  HelpText<"Pass <arg> to the offload compilers or the ones identified by -<triple>">,
+  MetaVarName<"<triple> <arg>">, Group<Link_Group>;
```
Can we make this a separate patch?

linehill replied:
Done: #170467.

jhuber6 commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -6990,6 +6990,12 @@ const ToolChain &Driver::getToolChain(const ArgList &Args,
         break;
       case llvm::Triple::spirv32:
       case llvm::Triple::spirv64:
+        if (Target.getOSName() == "hipspv" ||
```
I guess my concern is that we should not need to do string comparisons off of triple based things. Do we not have enum values for these?

pjaaskel replied:
@linehill can we just add a chipstar OS enum? I think it's a better name also than 'hipspv' now.

linehill replied:
Opened PR #170655 for adding the triple enums.

jhuber6 reviewed (commented):
Looking alright in general, some nits

linehill committed:
Add `-Xoffload-compiler` option

... to forward input to clang-linker-wrapper's device compiler
invocation.

linehill committed:
[Triple] Add "hipspv" and "chipstar" OS components

These are for Clang driver for selecting HIPSPV toolchain. The former
component sets up pre-existing HIPSPV toolchain behavior and the
latter tailors the toolchain for
[chipStar](https://github.com/CHIP-SPV/chipStar).

linehill committed:
HIPSPV: fixup. Use HIPSPV and ChipStar triple enums

linehill committed:
fixup: revert hipspv OS component

linehill committed:
HIPSPV: fixup: revert checks for --offload=spirv64-unknown-hipspv

linehill committed:
HIPSPV: fixup: revert code for "hipspv" triple OS component

linehill committed:
HIPSPV: add deprecation notice for --offload=spirv{32,64}

linehill committed:
fixup: forgot a test input

jhuber6 reviewed (commented):
Seems reasonable overall, I've accepted the dependencies so I'll review again once this is rebased

linehill committed:
fixup: --offload -> --offload-targets in ReleaseNotes.rst

jhuber6 committed:
Merge branch 'main' into hipspv-new-offload-driver

jhuber6 reviewed (approved):
Seems fine, thanks for dealing with this since I am pretty close to flipping the default once some downstream tests finish

jhuber6 commented:
> I'm preparing fixes in here: #178664.

I pushed https://github.com/llvm/llvm-project/commit/9ae6d8f565515a0199b18d6f156b1a41f328af1a, we'll see if the bots turn green. If not, I'll revert and we can track a new PR
